tools/meteor-services/deploy.js

```diff
@@ -321,6 +323,206 @@ function canonicalizeSite(site) {
+  if(firstRun) {
+    // Set up tracking variables to persist and be updated throughout polling
+    start = new Date().getTime();
+    pollingRounds = 1;
+    currentStatus = {buildStatus: null, isActive: false};
+    
+    // Timeouts must be done within Promises, so we set a dummy Promise to
+    // wait before initiating polling
+    return new Promise(function (resolve) {
+      let timer = setTimeout(() => resolve(null), initialWaitTimeMillis);
+    }).then(fiberHelpers.bindEnvironment(function() {
+      return pollForDeploy(pollingConfiguration, versionId, false);
+    })).await();
+  } else {
+    // Do a call to the version-status endpoint for the specified versionId
+    let versionStatusResult = deployRpc({
+      method: 'GET',
+      operation: 'version-status',
+      site: versionId,
+      expectPayload: [],
+      printDeployURL: false,
+    })
+    // We've done one call, so update the number of polling rounds
+    pollingRounds++;
+    let terminal = false;
+
+    //Check the details of the Version Status response and compare to last call
+    if(versionStatusResult &&
+      versionStatusResult.payload &&
+      versionStatusResult.payload.buildStatus) {
+      // Collect all of the statuses from the returned rpc call
+      const {buildStatus, deployStatus, isActive} = versionStatusResult.payload;
+
+      // If the build status has changed, we should report on that change to the user
+      // Allowable values for buildStatus are :
+      // ['missing', 'failed-temporarily', 'failed-permanently', 'success', 'building']
+      //    We will ignore the 'missing' buildStatus, and treat 'failed-temporarily'
+      //    the same as we do 'building', trusting the server to change the status.
+      if(currentStatus.buildStatus !== buildStatus) {
+        // Update our tracking variable for buildStatus
+        currentStatus.buildStatus = buildStatus;
+        // If we have reached a terminal buildStatus, handle it
+        if (terminalBuildStatuses.indexOf(buildStatus) >= 0) {
+          if(buildStatus == 'success') {
+            Console.info('Successfully built app image, initiating deploy');
+            // Reset the polling rounds to poll faster for a change in deployStatus
+            // Basically, we don't necessarily expect deploying to take a long time simply
+            // because building did.
+            pollingRounds = 1;
+          } else {
+            // Any other terminal status is failure, so report on it
+            Console.info(`Your app failed to build, with buildStatus: ${buildStatus}`);
+            // We also want to stop polling, since deploy cannot succeed with a failed build
+            terminal = true;
+          }
+        } else if (buildStatus !== 'building') {
+          // This will report if the buildStatus changes to a non-terminal, unexpected state
+          //    e.g. failed-temporarily
+          Console.info(`Build state changed to nonterminal state: ${buildStatus}`)
```

Comment:
I think exposing this lexicon to the user is probably not going to be as clear as we want it to be, especially since "terminal" might be confused with the terminal (a.k.a. shell, tty, console).  In fact, I think the _termin_-ology might be a bit too formal grammar, and possibly inappropriate, rather than just calling it `finished`.

Reply:
See also: my later suggestion about getting rid of the functional purpose of the `terminal` boolean entirely. 😄

Reply:
Agreed, I think we can just say `Build status: failed-temporarily` or whatever.

Note that galaxy-server (in AppVersionItem.jsx) has maps buildStatusMap and deployStatusMap which translate the codes into more human-suitable messages. Maybe we should extract those maps to a AppVersion.helper and send them down on the RPC (alongside the code) so we can have the terminology match the web interface.